File: server/app/services/google_auth.py
import json
import logging
import time
from urllib.parse import urlencode

import httpx
from sqlalchemy.orm import Session

GOOGLE_AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth"
GOOGLE_TOKEN_URL = "https://oauth2.googleapis.com/token"
SCOPES = [
    "https://www.googleapis.com/auth/calendar.readonly",
    "https://www.googleapis.com/auth/photoslibrary.readonly",
]

logger = logging.getLogger(__name__)


def build_auth_url(client_id: str, redirect_uri: str) -> str:
    params = {
        "client_id": client_id,
        "redirect_uri": redirect_uri,
        "response_type": "code",
        "scope": " ".join(SCOPES),
        "access_type": "offline",
        "prompt": "consent",
    }
    return f"{GOOGLE_AUTH_URL}?{urlencode(params)}"


async def exchange_code(
    code: str, client_id: str, client_secret: str, redirect_uri: str,
) -> dict:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            GOOGLE_TOKEN_URL,
            data={
                "code": code,
                "client_id": client_id,
                "client_secret": client_secret,
                "redirect_uri": redirect_uri,
                "grant_type": "authorization_code",
            },
        )
        response.raise_for_status()
        return response.json()


async def refresh_access_token(
    refresh_token: str, client_id: str, client_secret: str,
) -> dict:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            GOOGLE_TOKEN_URL,
            data={
                "refresh_token": refresh_token,
                "client_id": client_id,
                "client_secret": client_secret,
                "grant_type": "refresh_token",
            },
        )
        response.raise_for_status()
        return response.json()


async def get_valid_access_token(
    session: Session,
    encryption_key: bytes,
    client_id: str,
    client_secret: str,
) -> str | None:
    """Get a valid Google access token, refreshing if expired.

    Returns the access token string, or None if no integration exists
    or the token cannot be refreshed.
    """
    from server.app.models import Integration
    from server.app.services.encryption import decrypt, encrypt

    integration = session.query(Integration).filter(
        Integration.provider == "google", Integration.status == "connected"
    ).first()
    if not integration:
        return None

    try:
        tokens = json.loads(decrypt(integration.credentials, encryption_key))
    except Exception as e:
        logger.error(f"Failed to decrypt Google credentials: {e}")
        return None

    access_token = tokens.get("access_token")
    expires_at = tokens.get("expires_at")

    # Check if token is expired (or missing expiry = treat as expired)
    is_expired = expires_at is None or time.time() >= expires_at

    if not is_expired:
        return access_token

    # Token is expired, try to refresh
    refresh_token = tokens.get("refresh_token")
    if not refresh_token:
        logger.warning("Google access token expired and no refresh token available")
        return None

    try:
        new_tokens = await refresh_access_token(
            refresh_token=refresh_token,
            client_id=client_id,
            client_secret=client_secret,
        )
    except Exception as e:
        logger.error(f"Failed to refresh Google access token: {e}")
        return None

    # Update stored tokens
    tokens["access_token"] = new_tokens["access_token"]
    if "refresh_token" in new_tokens:
        tokens["refresh_token"] = new_tokens["refresh_token"]
    tokens["expires_at"] = time.time() + new_tokens.get("expires_in", 3600)

    integration.credentials = encrypt(json.dumps(tokens), encryption_key)
    session.commit()

    return new_tokens["access_token"]

File: src/server/services/google-auth.ts
import type Database from 'better-sqlite3'
import { getIntegrationByProvider, upsertIntegration } from '@server/db/queries/integrations.js'
import { decrypt, encrypt } from '@server/services/encryption.js'

const GOOGLE_AUTH_URL = 'https://accounts.google.com/o/oauth2/v2/auth'
const GOOGLE_TOKEN_URL = 'https://oauth2.googleapis.com/token'
const SCOPES = [
  'https://www.googleapis.com/auth/calendar.readonly',
  'https://www.googleapis.com/auth/photospicker.mediaitems.readonly',
]

export interface TokenResponse {
  access_token: string
  refresh_token?: string
  expires_in?: number
  token_type: string
}

/**
 * Build a Google OAuth authorization URL.
 */
export function buildAuthUrl(clientId: string, redirectUri: string): string {
  const params = new URLSearchParams({
    client_id: clientId,
    redirect_uri: redirectUri,
    response_type: 'code',
    scope: SCOPES.join(' '),
    access_type: 'offline',
    prompt: 'consent',
  })
  return `${GOOGLE_AUTH_URL}?${params.toString()}`
}

/**
 * Exchange an authorization code for tokens.
 */
export async function exchangeCode(
  code: string,
  clientId: string,
  clientSecret: string,
  redirectUri: string,
): Promise<TokenResponse> {
  const response = await fetch(GOOGLE_TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      code,
      client_id: clientId,
      client_secret: clientSecret,
      redirect_uri: redirectUri,
      grant_type: 'authorization_code',
    }),
  })

  if (!response.ok) {
    const errorBody = await response.text()
    throw new Error(`Token exchange failed: ${response.status} ${response.statusText} - ${errorBody}`)
  }

  return response.json() as Promise<TokenResponse>
}

/**
 * Refresh an access token using a refresh token.
 */
export async function refreshAccessToken(
  refreshToken: string,
  clientId: string,
  clientSecret: string,
): Promise<TokenResponse> {
  const response = await fetch(GOOGLE_TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      refresh_token: refreshToken,
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: 'refresh_token',
    }),
  })

  if (!response.ok) {
    const errorBody = await response.text()
    throw new Error(`Token refresh failed: ${response.status} ${response.statusText} - ${errorBody}`)
  }

  return response.json() as Promise<TokenResponse>
}

/**
 * Get a valid Google access token, refreshing if expired.
 * Returns the access token string, or null if no integration exists
 * or the token cannot be refreshed.
 */
export async function getValidAccessToken(
  db: Database.Database,
  encryptionKey: Buffer,
  clientId: string,
  clientSecret: string,
): Promise<string | null> {
  const integration = getIntegrationByProvider(db, 'google')
  if (!integration || integration.status !== 'connected') {
    return null
  }

  let tokens: Record<string, unknown>
  try {
    tokens = JSON.parse(decrypt(integration.credentials, encryptionKey))
  } catch {
    return null
  }

  const accessToken = tokens.access_token as string | undefined
  const expiresAt = tokens.expires_at as number | undefined

  // Check if token is expired (or missing expiry = treat as expired)
  const isExpired = expiresAt == null || Date.now() / 1000 >= expiresAt

  if (!isExpired) {
    return accessToken ?? null
  }

  // Token is expired, try to refresh
  const refreshToken = tokens.refresh_token as string | undefined
  if (!refreshToken) {
    return null
  }

  let newTokens: TokenResponse
  try {
    newTokens = await refreshAccessToken(refreshToken, clientId, clientSecret)
  } catch {
    return null
  }

  // Update stored tokens
  tokens.access_token = newTokens.access_token
  if (newTokens.refresh_token) {
    tokens.refresh_token = newTokens.refresh_token
  }
  tokens.expires_at = Date.now() / 1000 + (newTokens.expires_in ?? 3600)

  const encryptedCredentials = encrypt(JSON.stringify(tokens), encryptionKey)
  upsertIntegration(db, 'google', encryptedCredentials, 'connected')

  return newTokens.access_token
}
